### ISSUE TITLE
hook problems with killing pane

### ISSUE THREAD

**Issue by poetaman** [tmux/tmux#2848]
### Issue description
Let's say there is a session with a window and few split panes in that window. If one were to kill a pane with ``prefix-x`` (tied to ``kill-pane`` command) then hook after-kill-pane is triggered (though with empty ``hook_pane``). If one were to kill a pane by entering ``exit<CR>`` then nothing is triggered.

How can this be improved?
1. after-kill-pane should be triggered even if a pane is killed using ``exit<CR>``
2. after-kill-pane should have the hook_pane option equal to the pane that was killed
3. pane-focus-in should be triggered anytime focus moves to a new pane because another pane was killed. Note: pane-focus-in is always triggered if a window is killed, so not sure why not if another pane is killed.

### Required information

Please provide the following information:

* tmux version (`tmux -V`) next-3.3.
* Platform (`uname -sp`) Darwin arm.
* $TERM inside and outside of tmux (`echo $TERM`) xterm-256color, tmux-256color.

**Comment by nicm** [comment]
> If one were to kill a pane by entering exit<CR> then nothing is triggered.

Apart from `pane-exited` or `pane-died`? Why don't you look at the list of hooks in the man page?

**Comment by reportaman** [comment]
@nicm I see, given after-kill-pane did trigger on some cases I did not bother looking for another hook that does the same thing... Will test now. Maybe after-pane-exited is deprecated and should be removed???

**Comment by nicm** [comment]
Why would `after-kill-pane` ever trigger when you have not used the `kill-pane` command?

**Comment by reportaman** [comment]
@nicm I see what you mean, I thought exiting effectively triggers``kill-pane`` but not so much. I now understand that those hooks only if those commands are run. I think kill-pane is special in that a pane can be killed without a command, thus the apparently wrong expectation from it...

**Comment by nicm** [comment]
No `after` hooks are for commands.

I know it is a pain that there are multiple hooks for basically the same thing in slightly different ways, maybe we should consolidate them with eg a hook that always fires when the pane closes no matter how it happens.

**Comment by reportaman** [comment]
@nicm ``pane-exited`` gives the correct ``hook_pane`` id of pane that exited. ``after-kill-pane`` does not, ``hook_pane`` is empty for it.

**Comment by reportaman** [comment]
Also, when one creates pane with ``split-window -h`` or ``split-window -v`` events ``pane-focus-out`` and ``pane-focus-in`` are triggered. But when exits the pane with ``exit<CR>`` there is no ``pane-focus-in``. While one can look at the current ``#{pane_id}`` during ``pane-exited`` hook, not sure if you intended to also send ``pane-focus-in`` but are missing it, just informing...

**Comment by nicm** [comment]
What are the actions for this issue? Is there a bug or what change would you like to see?

**Comment by reportaman** [comment]
@nicm I tested again, these are some points you could consider:

1) after-kill-pane's hook_pane is empty, perhaps this can be the pane_id of the pane that just got killed.

2) after-kill-pane is triggered after pane-focus-in, ideally it should be other way round. That will be consistent with the focus ordering that is followed elsewhere, example: pane-exited is triggered before pane-focus-in.

One thing you have implemented is that pane-focus-in is triggered after pane-exited, thanks for that! I think that was fixed In commit for https://github.com/tmux/tmux/issues/2966

**Comment by vlappa** [comment]
I can confirm that `after-kill-pane`, `hook_pane` is empty, while with `pane-exited`, `hook_pane` is not empty

**Comment by vlappa** [comment]
I can confirm that `after-kill-pane`, `hook_pane` is empty, while with `pane-exited`, `hook_pane` is not empty.

`after-kill-pane` is the functionality I'm looking for (`pane-exited` and `pane-died` doesn't seem to do anything at `kill-pane`), but I also need the `hook_pane`

**Comment by vlappa** [comment]
What I want to do is: manage a tmux session in the following way:

If a application / process dies in a pane, we assume the user has just stopped it temporarily of that it exited unexpectedly, we assume it's still part of the session. But if the user kills the pane, the process / application is removed from the session.

So we need to know which pane belongs to which process (the application should probably tell us the pane_id / TMUX_PANE) and if a pane is killed, we've to know it's pane_id, so we can remove the application / process running in it, from our 'session'. 

So I think we need `after-kill-pane` to give us the `hook_pane`, we can't use `pane_id` here, cause it will send the current `pane_id`, not the one that just was killed.

**Comment by vlappa** [comment]
I'm able to work around the missing `pane_hook` / `pane_id`, by listing the pane id's of a particular session after I receive a `after-kill-pane` (and `window-unlinked`) message and if the reported pane of the application is not listed, I remove the application from the session (see my previous comment). 

Receiving the `pane_hook` after `after-kill-pane` would make things more simple (all though I still have to handle `window-unlined` probably).

**Comment by nicm** [comment]
Yeah... we don't add these for `after-` hooks. Please try this: [after_hook.diff.txt](https://github.com/user-attachments/files/23257242/after_hook.diff.txt)
